### DISCUSSION TITLE
Importance of ordered listeners

### DISCUSSION THREAD

**Discussion by drieseng** [quartznet/quartznet#1384] [General]
To ensure we notify listeners in the same order in which they were added, we currently use a custom **OrderedDictionary\<TKey, TValue>** in **ListenerManagerImpl** and a **List\<T>** for **ISchedulerListener** instances in both **ListenerManagerImpl** and **QuartzScheduler**.

We're not always consistent in this, since we use a regular **ConcurrentDictionary\<TKey, TValue>** in **QuartzScheduler** for the ("internal") job and trigger listeners.

I'm currently looking at improving performance, reducing contention and reducing allocations in this area.
I wonder how important it is that we notiffy the listeners in the same order in which they were added.
For internal listeners, this appears not be this important or at least less important than reduced contention (since we use **ConcurrentDictionary\<TKey, TValue>**).

@lahma How important is it to invoke global listeners in the order they were added? Would it be ok to replace the use of **OrderedDictionary<TKey,TValue>** in **ListenerManagerImpl** with **ConcurrentDictionary<TKey,TValue>**?

An alternate solution for those users - like myself - that do not have a need to register global listeners, would be to allow one to specify a custom (no-op) **IListenerManager** implementation. For now, we can just introduce a new overload of `DirectSchedulerFactory.CreateScheduler(...)` that takes an extra **IListenerManager** argument. In the long run, a builder for **IScheduler**/**StdScheduler** would perhaps be a better solution.

**Comment by drieseng** [comment]
@lahma Can you have a look at the questions above?

Summary:
* Is it important to notify listeners in the order in which they are added/registered?
* Would you accept a PR that allows one to specify a custom **IListenerManager** implementation?

**Reply by lahma** [reply]
> * Is it important to notify listeners in the order in which they are added/registered?

I think so yes. People may have opted to specific ordering and a chain of tasks to take take place in correct order. 

> * Would you accept a PR that allows one to specify a custom **IListenerManager** implementation?

I think this would be more easily achieved if Quartz would have hard dependency on MS DI library and users are allowed to register alternate implementations. This is what I'd like see to be in the 4.0 roadmap.

**Reply by drieseng** [reply]
> > * Would you accept a PR that allows one to specify a custom **IListenerManager** implementation?
> 
> I think this would be more easily achieved if Quartz would have hard dependency on MS DI library and users are allowed to register alternate implementations. This is what I'd like see to be in the 4.0 roadmap.

To do this, you'd still need to make it possible to somehow inject an **IListenerManager** instance.

[DRAFT]
I propose to:
* Add a `IListenerManager ListenerManager { get; set; }` property to **QuartzSchedulerResources**.
* Add a `CreateScheduler(...)` overload to *DirectSchedulerFactory* that takes a **IListenerManager** argument.

I'll have a closer look later today.

**Reply by lahma** [reply]
For backwards compatibility I think it should be possible to define a property like `quartz.listenerManager.type` = 'MyCustomType'. This will then allow also using DI services (which I'm currently prototyping while I have some free time). `DirectSchedulerFactory`is a bit different beast, I don't know if something like `SchedulerOptions` would make sense (I'm trying to push API towards `IOptions` kind of naming when setup code needs to be pushed via constructors.
